Frameworks/Foundation/NSScanner.mm

```diff
@@ -575,118 +557,283 @@ - (BOOL)scanUpToString:(NSString*)string intoString:(NSString**)intoString {
 /**
  @Status Interoperable
 */
-- (BOOL)scanCharactersFromSet:(id)charset intoString:(NSString**)stringp {
-    unsigned int length = (unsigned int)[_string length];
-    unichar* result = (unichar*)IwMalloc(length * sizeof(unichar));
-    int resultLength = 0;
-    BOOL scanStarted = NO;
+- (BOOL)scanCharactersFromSet:(NSCharacterSet* _Nonnull)charset intoString:(NSString* _Nullable*)stringp {
+    const NSUInteger length = _string.length;
+    NSUInteger i = [self _indexOfNextUnskippedCharacter];
 
-    for (; _location < length; _location++) {
-        unichar unicode = (unichar)[_string characterAtIndex:_location];
+    std::vector<unichar> result;
+    result.reserve(length - i);
+    for (; i < length; ++i) {
+        unichar unicode = (unichar)[_string characterAtIndex:i];
 
-        if ([_skipSet characterIsMember:unicode] && (scanStarted == NO)) {
-            // do nothing
+        if ([charset characterIsMember:unicode]) {
```

Comment:
these two seem close enough to be mergeable, if you abstract out the terminal condition